src/H5Dint.c

```diff
@@ -947,8 +947,8 @@ H5D__update_oh_info(H5F_t *file, H5D_t *dset, hid_t dapl_id)
         HGOTO_ERROR(H5E_DATASET, H5E_CANTPIN, FAIL, "unable to pin dataset object header");
 
     /* Check for creating dataset with unusual datatype */
-    if (H5T_is_numeric_with_unusual_unused_bits(type) &&
-        !(H5O_has_chksum(oh) || (H5F_RFIC_FLAGS(file) & H5F_RFIC_UNUSUAL_NUM_UNUSED_NUMERIC_BITS)))
+    if (!(H5O_has_chksum(oh) || (H5F_RFIC_FLAGS(file) & H5F_RFIC_UNUSUAL_NUM_UNUSED_NUMERIC_BITS)) &&
```

Comment:
Per @fortnern's comment, invert a couple of these checks to do the easy work first, then call `H5T_is_numeric_with_unusual_unused_bits`